gdc_client/parcel/segment.py

```diff
@@ -320,9 +346,19 @@ def finish_download(self):
     def wait_for_completion(self):
         try:
             since_save = 0
-            while not self.is_complete():
+            num_tasks_completed = 0
+            while num_tasks_completed != self.total_tasks:
                 while since_save < self.save_interval:
                     interval = self.q_complete.get()
+                    # Once a process completes a tasks (sucess or failure),
+                    # it will return a sentinal value (None) to main process
+                    # to indicate that a task was completed
+                    if interval is None:
+                        num_tasks_completed += 1
+                        if num_tasks_completed == self.total_tasks:
+                            break
+                        continue
```

Comment:
Can we use the `data` attribute in the `Interval` class to signal success or failure instead of adding an extra sentinel?

Reply:
the producer is already using None as a sentinel value to signal to the worker processes to exit.  Seems functionally the same to have to create a dummy Interval with just a key set in the data attribute and then check that in segment.py to increment the number of tasks completed. if we wanted to take this route, would probably want to update the None being sent as a sentinel to the worker processes to exit with the dummy interval to keep things in sync.

Reply:
and that might not work. if we add a "success" key in the data dict, this would determine whether the small interval chunk of the larger Interval given to a worker process was completed successfully or not, not whether or not the whole Interval was successful or not.